internal/runtime/client/client.go

```diff
@@ -74,180 +92,247 @@ func (c *client) Extension(ext *runtimev1.Extension) ExtensionClient {
-	Invoke(ctx context.Context, in, out runtime.Object) error
+func (e *extensionClient) Discover() (*runtimev1.Extension, error) {
+	gvh, err := e.client.catalog.GroupVersionHook(hooksv1alpha1.Discovery)
+	if err != nil {
+		return nil, err
+	}
+
+	request := &hooksv1alpha1.DiscoveryHookRequest{}
+	response := &hooksv1alpha1.DiscoveryHookResponse{}
+
+	// Future work: The discovery runtime extension could be operating on a different hook version than
+	// the latest. We will have to loop through different versions of the discover hook here to actually
+	// finish discovery.
+	opts := &httpCallOptions{
+		catalog: *e.client.catalog,
+		config:  e.ext.Spec.ClientConfig,
+		gvh:     gvh,
+	}
+	if err := httpCall(context.Background(), request, response, opts); err != nil {
+		return nil, errors.Wrap(err, "failed to call the discovery extension")
+	}
+
+	modifiedExtension := &runtimev1.Extension{}
+	e.ext.DeepCopyInto(modifiedExtension)
+	for _, extension := range response.Extensions {
+		modifiedExtension.Status.RuntimeExtensions = append(
+			modifiedExtension.Status.RuntimeExtensions,
+			runtimev1.RuntimeExtension{
+				Name:           extension.Name + "." + e.ext.Name,
+				Hook:           runtimev1.Hook(extension.Hook),
+				TimeoutSeconds: extension.TimeoutSeconds,
+				FailurePolicy:  (*runtimev1.FailurePolicyType)(extension.FailurePolicy),
+			},
+		)
+	}
+
+	// TODO: Decide if we also want to register inside this function.
+
+	return modifiedExtension, nil
 }
 
-type serviceClient struct {
-	client *client
-	svc    catalog.Hook
-	opts   []ServiceOption
+func (e *extensionClient) Register() error {
+	return e.client.registry.Add(e.ext)
 }
 
-var _ ServiceClient = &serviceClient{}
+func (e *extensionClient) Unregister() error {
+	return e.client.registry.Remove(e.ext)
+}
 
-func (c *client) ServiceOld(svc catalog.Hook, opts ...ServiceOption) ServiceClient {
-	return &serviceClient{
-		client: c,
-		svc:    svc,
-		opts:   opts,
-	}
+type httpCallOptions struct {
+	// TODO: Left catalog as not a pointer as the zero value of a pointer type is
+	// not acceptable in the code. The code doest not work if the catalog is nil.
+	catalog       catalog.Catalog
+	config        runtimev1.ExtensionClientConfig
+	gvh           catalog.GroupVersionHook
+	name          string
+	timeout       *int32
+	failurePolicy *runtimev1.FailurePolicyType
 }
 
-func (s *serviceClient) Invoke(ctx context.Context, in, out runtime.Object) error {
-	serviceOpts := &ServiceOptions{}
-	for _, o := range s.opts {
-		o.ApplyToServiceOptions(serviceOpts)
-	}
+func httpCall(ctx context.Context, request, response runtime.Object, opts *httpCallOptions) error {
 
-	gvh, err := s.client.catalog.GroupVersionHook(s.svc)
+	url, err := urlForExtension(opts.config, opts.gvh, opts.name)
 	if err != nil {
 		return err
 	}
 
-	requireConversion := serviceOpts.Version != gvh.Version
+	requireConversion := opts.gvh.Version != request.GetObjectKind().GroupVersionKind().Version
```

Comment:
Let's add a TODO.

I think eventually we have to resolve opts.gvh to the corresponding request/response gvk. Then we have to decide for request and response individually if conversion is needed.

This shouldn't change a lot but it feels cleaner to me to diff the version of request / response individually vs relying on having the same version in the hook and in request/response (which is probably always true).